aten/src/ATen/native/Pooling.cpp

```diff
@@ -1,16 +1,177 @@
 #include <ATen/ATen.h>
 
+#include <ATen/Dispatch.h>
+#include <ATen/NamedTensorUtils.h>
 #include <ATen/NativeFunctions.h>
 #include <ATen/TensorUtils.h>
-#include <ATen/NamedTensorUtils.h>
+#include <ATen/native/Pool.h>
 #include <ATen/native/xnnpack/Engine.h>
 #include <c10/macros/Macros.h>
 #include <c10/util/Exception.h>
 
+#include <cmath>
+#include <cstring>
+#include <limits>
 #include <tuple>
+#include "ATen/Parallel.h"
+#include "Functions.h"
 
 namespace at { namespace native {
 
+namespace {
+
+template <typename scalar_t>
+static void max_pool2d_out_impl(
+    const scalar_t* __restrict IP,
+    scalar_t* __restrict OP,
+    const int64_t NB,
+    const int64_t NC,
+    const int64_t IH,
+    const int64_t IW,
+    const int64_t OH,
+    const int64_t OW,
+    const int64_t KH,
+    const int64_t KW,
+    const int64_t SI,
+    const int64_t SJ,
+    const int64_t PI,
+    const int64_t PJ,
+    const int64_t DI,
+    const int64_t DJ) {
+  // Value to fill the padded region with
+  constexpr auto FILL = std::numeric_limits<scalar_t>::lowest();
+
+  // Because of padding, the kernel index may be off bounds. This computes the
+  // offset to place the kernel index in bounds.
+  const auto ker_offset = [&](const int64_t excess) {
+    return std::max(static_cast<int64_t>(std::ceil(excess / (DI * -1.0))), 0L);
+  };
+
+  // For each output row
+  at::parallel_for(
+      0, NB * NC * OH, 0, [&](const int64_t begin, const int64_t end) {
+        for (int64_t it = begin; it < end; ++it) {
+          const auto tensor_idx = it / OH;
+          const auto oi = it % OH;
+
+          // Compute kernel first and last rows within input bounds
+          const auto kis = ker_offset(oi * SI - PI);
+          const auto kie =
+              KH - ker_offset(((IH - 1) - (oi * SI + (KH - 1) * DI - PI)));
+
+          // Input tensor offset pointing at first element of current frame
+          const auto in_offset = tensor_idx * IH * IW;
+
+          // Allocate a buffer for the current input row to apply max
+          // column-wise
+          std::vector<scalar_t> buffer(IW + (PJ << 1), FILL);
+
+          // Copy first valid (non-padded) row to skip one loop iteration
+          const auto copy_offset = in_offset + (oi * SI + kis * DI - PI) * IW;
+          std::copy(
+              IP + copy_offset, IP + copy_offset + IW, buffer.begin() + PJ);
+
+          // Compute max column-wise for current output row
+          for (int64_t ki = kis + 1; ki < kie; ++ki) {
+            const auto ii = oi * SI + ki * DI - PI;
+            for (int64_t ij = 0; ij < IW; ++ij) {
+              const auto val = IP[in_offset + ii * IW + ij];
+              buffer[ij + PJ] =
+                  std::isnan(val) ? val : std::max(buffer[ij + PJ], val);
+            }
+          }
+
+          // Compute max for each cell in the current output row
+          const auto out_offset = (tensor_idx * OH + oi) * OW;
```

Comment:
tensor_idx * OH  == it I think, you can use it, and add comment to clearify what 'it' is. Use int64_t instead as well.